[BUG] We need to validate the existence of the file and correct data structure of the token file

### Description

We need to validate the existence of the file and correct data structure of the token file provided by the user with `credentials_file`

It looks like [currently](https://github.com/civo/terraform-provider-civo/commit/cbcaa08ce32594b41dbe54061781150131e1b91e#diff-99691afc1cd01012134ee4d109283a9b3e1d26ebf58dde8a5c53693cbb185fd8R158) we are not validating the data or the existence of the `credentials_file` provided by the user.



### Acceptance Criteria

* Ensure that we check for the existence of the file
* Validity of the file
* If the file is too large, over 20 megs or something like that, perhaps we should fail too even before trying to read the file to avoid issues. Not sure how this is handled right now --> This is a bonus and could be done as other ticket for all files in general.

Hi @uzaxirr,

For this issue, would the following approach suffice?

```Go
func readTokenFromFile(path string) (string, error) {
	// check if the file exists
	_, err := os.Stat(path)
	if errors.Is(err, fs.ErrNotExist) {
		return "", fmt.Errorf("file does not exist: %s", path)
	}

	// Check file size: 20 MB limit
	if err = utils.CheckFileSize(path, 20*1024*1024); err != nil {
		return "", err
	}

	// read the file
	data, err := os.ReadFile(path)
	if err != nil {
		return "", fmt.Errorf("error reading file (%s): %v", path, err)
	}

	var config struct {
		APIKeys map[string]string `json:"apikeys"`
		Meta    struct {
			CurrentAPIKey string `json:"current_apikey"`
		} `json:"meta"`
	}

	if err := json.Unmarshal(data, &config); err != nil {
		return "", fmt.Errorf("failed to parse JSON from %s: %v. Please ensure the input JSON is correctly formatted and all required fields are present", path, err)
	}

	if config.APIKeys == nil || config.Meta.CurrentAPIKey == "" {
		return "", fmt.Errorf("invalid structure in %s, missing required fields", path)
	}

	currentKeyName := config.Meta.CurrentAPIKey

	token, ok := config.APIKeys[currentKeyName]
	if !ok {
		return "", fmt.Errorf("API key '%s' not found in %s", currentKeyName, path)
	}
	return token, nil
}
```

I have question, should I return these errors to users, or just log it like below?
```Go
log.Printf("[ERROR] error reading token from CLI config: %v\n", err)
```

And if we are to return these errors, what to do in the following scenario?

If I encountered an error while reading from `credentials_file` and we fall back to `~/.civo.json` but got one error here as well. Should we club these two errors and show both?

<img width="661" alt="Screenshot 2024-08-09 185835" src="https://github.com/user-attachments/assets/d657fe9e-34e9-49dc-8e75-d35fed2d1bc8">

--- a/civo/provider.go
+++ b/civo/provider.go
@@ -23,9 +23,11 @@ import (
 	"github.com/civo/terraform-provider-civo/civo/size"
 	"github.com/civo/terraform-provider-civo/civo/ssh"
 	"github.com/civo/terraform-provider-civo/civo/volume"
+	"github.com/civo/terraform-provider-civo/internal/utils"
 	"github.com/hashicorp/go-cty/cty"
 	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
 	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	"github.com/mitchellh/go-homedir"
 )
 
 var (
@@ -121,11 +123,11 @@ func providerConfigure(d *schema.ResourceData) (interface{}, error) {
 		regionValue = region.(string)
 	}
 
-	if token, ok, source := getToken(d); ok {
+	if token, source, err := getToken(d); err == nil {
 		tokenValue = token.(string)
 		tokenSource = source
 	} else {
-		return nil, fmt.Errorf("[ERR] No token configuration found in $CIVO_TOKEN or ~/.civo.json. Please go to https://dashboard.civo.com/security to fetch one")
+		return nil, fmt.Errorf("[ERR] No token configuration found in $CIVO_TOKEN or credentials_file or ~/.civo.json. Please go to https://dashboard.civo.com/security to fetch one: %v", err)
 	}
 
 	if apiEndpoint, ok := d.GetOk("api_endpoint"); ok {
@@ -160,48 +162,50 @@ func providerConfigure(d *schema.ResourceData) (interface{}, error) {
 	return client, nil
 }
 
-func getToken(d *schema.ResourceData) (interface{}, bool, string) {
-	var exists = true
+func getToken(d *schema.ResourceData) (interface{}, string, error) {
 
 	// Gets you the token atrribute value or falls back to reading CIVO_TOKEN environment variable
 	if token, ok := d.GetOk("token"); ok {
-		return token, exists, "environment variable"
+		return token, "environment variable", nil
 	}
 
 	// Check for credentials file specified in provider config
 	if credFile, ok := d.GetOk("credentials_file"); ok {
-		token, err := readTokenFromFile(credFile.(string))
-		if err == nil {
-			return token, exists, "credentials file"
+		path, err := homedir.Expand(credFile.(string))
+		if err != nil {
+			return nil, "", fmt.Errorf("error expanding %v: %w", credFile, err)
 		}
+		token, err := readTokenFromFile(path)
+		if err == nil {
+			return token, "credentials file", nil
+		}
+		return nil, "", fmt.Errorf("error reading from credentials_file: %v", err)
 	}
 
 	// Check for default CLI config file
-	homeDir, err := getHomeDir()
+	homeDir, err := homedir.Dir()
 	if err == nil {
 		token, err := readTokenFromFile(filepath.Join(homeDir, ".civo.json"))
 		if err == nil {
-			return token, exists, "CLI config file"
+			return token, "CLI config file", nil
 		}
+		return nil, "", fmt.Errorf("error reading from ~/.civo.json: %v", err)
 	}
 
-	return nil, !exists, ""
+	return nil, "", err
 
-}
-
-var getHomeDir = func() (string, error) {
-	if home := os.Getenv("HOME"); home != "" {
-		return home, nil
-	}
-	// Fall back to os.UserHomeDir() if HOME is not set
-	return os.UserHomeDir()
 }
 
 func readTokenFromFile(path string) (string, error) {
-	data, err := os.ReadFile(path)
-
-	if err != nil {
+	// Check file size: 20 MB limit
+	if err := utils.CheckFileSize(path); err != nil {
 		return "", err
+	}
+
+	// read the file
+	data, err := os.ReadFile(path)
+	if err != nil {
+		return "", fmt.Errorf("error reading file (%s): %w", path, err)
 	}
 
 	var config struct {
@@ -211,8 +215,22 @@ func readTokenFromFile(path string) (string, error) {
 		} `json:"meta"`
 	}
 
+	exampleJSON := `
+{
+	"apikeys": {
+		"tf_key": "token_here"
+	},
+	"meta": {
+		"current_apikey": "tf_key"
+	}
+}`
+
 	if err := json.Unmarshal(data, &config); err != nil {
-		return "", err
+		return "", fmt.Errorf("failed to parse JSON from '%s': %w. Please ensure the input JSON file is correctly formatted and all required fields are present. Expected format:\n%s", path, err, exampleJSON)
+	}
+
+	if config.APIKeys == nil || config.Meta.CurrentAPIKey == "" {
+		return "", fmt.Errorf("invalid structure in '%s', missing required fields. Expected format:\n%s", path, exampleJSON)
 	}
 
 	// Get the current API key name
@@ -220,11 +238,9 @@ func readTokenFromFile(path string) (string, error) {
 
 	// Fetch the corresponding token
 	token, ok := config.APIKeys[currentKeyName]
-
 	if !ok {
-		return "", fmt.Errorf("API key '%s' not found", currentKeyName)
+		return "", fmt.Errorf("API key '%s' not found in '%s'", currentKeyName, path)
 	}
-
 	return token, nil
 }
 

--- a/go.mod
+++ b/go.mod
@@ -4,7 +4,6 @@ require (
 	github.com/civo/civogo v0.3.73
 	github.com/google/uuid v1.3.1
 	github.com/hashicorp/go-cty v1.4.1-0.20200414143053-d3edf31b6320
-	github.com/hashicorp/go-version v1.7.0
 	github.com/hashicorp/terraform-plugin-sdk/v2 v2.31.0
 	github.com/stretchr/testify v1.8.4
 	golang.org/x/crypto v0.25.0
@@ -31,6 +30,7 @@ require (
 	github.com/hashicorp/go-multierror v1.1.1 // indirect
 	github.com/hashicorp/go-plugin v1.6.0 // indirect
 	github.com/hashicorp/go-uuid v1.0.3 // indirect
+	github.com/hashicorp/go-version v1.6.0
 	github.com/hashicorp/hc-install v0.6.2 // indirect
 	github.com/hashicorp/hcl/v2 v2.19.1 // indirect
 	github.com/hashicorp/logutils v1.0.0 // indirect
@@ -45,6 +45,7 @@ require (
 	github.com/mattn/go-colorable v0.1.12 // indirect
 	github.com/mattn/go-isatty v0.0.17 // indirect
 	github.com/mitchellh/copystructure v1.2.0 // indirect
+	github.com/mitchellh/go-homedir v1.1.0
 	github.com/mitchellh/go-testing-interface v1.14.1 // indirect
 	github.com/mitchellh/go-wordwrap v1.0.1 // indirect
 	github.com/mitchellh/mapstructure v1.5.0 // indirect

--- a/internal/utils/utils.go
+++ b/internal/utils/utils.go
@@ -5,8 +5,11 @@ package utils
 
 import (
 	"encoding/json"
+	"errors"
 	"fmt"
+	"io/fs"
 	"log"
+	"os"
 	"os/exec"
 	"regexp"
 	"sort"
@@ -20,6 +23,11 @@ import (
 	"github.com/hashicorp/go-cty/cty"
 	"github.com/hashicorp/go-version"
 	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+)
+
+const (
+	// FileSizeLimit limits the size of file to be used by the user
+	FileSizeLimit = int64(20 * 1024 * 1024) // 20 MB
 )
 
 // VersionInfo stores Provider's version Info
@@ -316,4 +324,19 @@ func ValidateUUID(v interface{}, k string) (ws []string, errors []error) {
 		errors = append(errors, fmt.Errorf("%q must be a valid UUID", k))
 	}
 	return
+}
+
+// CheckFileSize function checks if the file the file size is less than the allowed limit(current: 20MB)
+func CheckFileSize(path string) error {
+	fileInfo, err := os.Stat(path)
+	if err != nil {
+		if errors.Is(err, fs.ErrNotExist) {
+			return fmt.Errorf("file does not exist: %s", path)
+		}
+		return fmt.Errorf("error getting the file info: %w", err)
+	}
+	if fileInfo.Size() > FileSizeLimit {
+		return fmt.Errorf("file size exceeds the allowed limit of %d bytes", FileSizeLimit)
+	}
+	return nil
 }

